File: Top-Interview-Questions/Trees-and-Graphs-Medium-Collection/103-Binary-Tree-Zigzag-Level-Order-Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 // Solution #1 Breadth First Search
 // TC: O(number of nodes in the tree)
 // SC: O(max width of the tree)
 // Runtime: 0 ms, faster than 100.00%
 // Memory Usage: 12.2 MB, less than 75.89%
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        
        if(!root)
            return {};
        
        vector<vector<int>> res;
        
        queue<TreeNode*> q;
        q.push(root);
        
        int length;
        int zigzag = 0;
        
        while(!q.empty()){
            
            length = q.size();
            vector<int> level;
            
            for(int i = 0; i < length; i++){
                TreeNode *node = q.front();
                q.pop();
                level.push_back(node->val);
                
                if(node->left)
                    q.push(node->left);
                if(node->right)
                    q.push(node->right);
            }
            
            if(zigzag++%2){
                reverse(level.begin(), level.end());
                res.push_back(level);
            }else{
                res.push_back(level);
            }
        }
        
        
        return res;
    }
};

File: Top-Interview-Questions/Trees-and-Graphs-Medium-Collection/103-binary-tree-zigzag-level-order-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 * Runtime: 76 ms, faster than 41.47% of JavaScript online submissions for Binary Tree Zigzag Level Order Traversal.
 * Memory Usage: 37.3 MB, less than 6.82% of JavaScript online submissions for Binary Tree Zigzag Level Order Traversal.
 * TC: O(N), where N is number of elements in root
 * SC: O(logN) ~ O(N), where N is number of elements in root
 */
var zigzagLevelOrder = function(root) {
    const height = getHeight(root);
    const result = new Array(height).fill().map(() => []);

    traverse(root, result, 0);

    return result;
};

function traverse (node, result, depth) {
    if (!node) {
        return;
    }

    if (depth % 2 === 0) {
        result[depth].push(node.val);
    } else {
        result[depth].unshift(node.val);
    }

    traverse(node.left, result, depth + 1);
    traverse(node.right, result, depth + 1);
}

function getHeight (node) {
    if (!node) {
        return 0;
    }

    return Math.max(getHeight(node.left), getHeight(node.right)) + 1;
}